File: 2020/csharp/Aoc2020.Tests/Day07Tests.cs
using System.Linq;
using System.Collections.Generic;
using Xunit;

namespace Aoc2020.Tests
{
    public class Day07Tests
    {
        [Fact]
        public void ParseWorks()
        {
            var rules = BagRules.Parse(Utils.ReadInputLines("07-test"));
            Assert.True(rules.ContainsColor(new("light red")));
            Assert.True(rules.ContainsColor(new("dark orange")));
            Assert.True(rules.ContainsColor(new("bright white")));
            Assert.True(rules.ContainsColor(new("muted yellow")));
            Assert.True(rules.ContainsColor(new("shiny gold")));
            Assert.True(rules.ContainsColor(new("dark olive")));
            Assert.True(rules.ContainsColor(new("vibrant plum")));
            Assert.True(rules.ContainsColor(new("faded blue")));
            Assert.True(rules.ContainsColor(new("dotted black")));

            Assert.False(rules.ContainsColor(new("")));
            Assert.False(rules.ContainsColor(new("foo bar")));

            Assert.Equal(1, rules.SubbagCount(new("light red"), new("bright white")));
            Assert.Equal(2, rules.SubbagCount(new("light red"), new("muted yellow")));
            Assert.Equal(3, rules.SubbagCount(new("dark orange"), new("bright white")));
            Assert.Equal(4, rules.SubbagCount(new("dark orange"), new("muted yellow")));
            Assert.Equal(1, rules.SubbagCount(new("bright white"), new("shiny gold")));
            Assert.Equal(2, rules.SubbagCount(new("muted yellow"), new("shiny gold")));
            Assert.Equal(9, rules.SubbagCount(new("muted yellow"), new("faded blue")));
            Assert.Equal(1, rules.SubbagCount(new("shiny gold"), new("dark olive")));
            Assert.Equal(2, rules.SubbagCount(new("shiny gold"), new("vibrant plum")));
            Assert.Equal(3, rules.SubbagCount(new("dark olive"), new("faded blue")));
            Assert.Equal(4, rules.SubbagCount(new("dark olive"), new("dotted black")));
            Assert.Equal(5, rules.SubbagCount(new("vibrant plum"), new("faded blue")));
            Assert.Equal(6, rules.SubbagCount(new("vibrant plum"), new("dotted black")));
            Assert.Equal(0, rules.SubbagCount(new("faded blue"), new("dotted black")));
            Assert.Equal(0, rules.SubbagCount(new("dotted black"), new("foo bar")));
        }

        [Fact]
        public void BagsToHoldWorks()
        {
            var rules = BagRules.Parse(Utils.ReadInputLines("07-test"));
            Assert.Equal(4, rules.BagsToHold(new("shiny gold")).Length);
        }

        [Fact]
        public void TotalBagCountWorks()
        {
            var rules = BagRules.Parse(Utils.ReadInputLines("07-test"));
            Assert.Equal(32, rules.TotalSubbagCount(new("shiny gold")));
        }

        [Fact]
        public void Stars()
        {
            var run = new Day07();
            Assert.Equal("115", run.Result1());
            Assert.Equal("1250", run.Result2());
        }
    }
}

File: 2015/fsharp/Aoc2015.Tests/Day07.Tests.fs
module Day07.Tests

open Xunit
open Utils
open Day07

[<Fact>]
let ``Day07 part 1 works`` () =
    let circuit =
        Circuit.empty
        |> Circuit.add "123 -> x"
        |> Circuit.add "456 -> y"
        |> Circuit.add "x AND y -> d"
        |> Circuit.add "x OR y -> e"
        |> Circuit.add "x LSHIFT 2 -> f"
        |> Circuit.add "y RSHIFT 2 -> g"
        |> Circuit.add "NOT x -> h"
        |> Circuit.add "NOT y -> i"
        |> Circuit.run

    Assert.Equal(72us, circuit |> Circuit.get "d")
    Assert.Equal(507us, circuit |> Circuit.get "e")
    Assert.Equal(492us, circuit |> Circuit.get "f")
    Assert.Equal(114us, circuit |> Circuit.get "g")
    Assert.Equal(65412us, circuit |> Circuit.get "h")
    Assert.Equal(65079us, circuit |> Circuit.get "i")
    Assert.Equal(123us, circuit |> Circuit.get "x")
    Assert.Equal(456us, circuit |> Circuit.get "y")

[<Fact>]
let ``Day07 Stars`` () =
    try
        Assert.Equal("3176", job1 ())
        Assert.Equal("14710", job2 ())
    with :? System.NotImplementedException ->
        ()
